shell: add a directory for backends

jakub-uC opened the pull request:
This change organizes the file structure for the shell.
Files implementing backends are moved to a separate folder.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

nordic-krch commented on subsys/shell/CMakeLists.txt:
```diff
@@ -50,11 +31,6 @@ zephyr_sources_ifdef(
 )
 
-zephyr_sources_ifdef(
-  CONFIG_SHELL_LOG_BACKEND
```
i think that shell log backend shall stay in main dir. It's not a shell backend

rlubos commented on subsys/shell/backends/shell_telnet.c:
```diff
@@ -13,7 +13,7 @@
 #include <shell/shell_telnet.h>
 
-#include "shell_telnet_protocol.h"
+#include "../shell_telnet_protocol.h"
```
I think you should move corresponding private header as well, it's only used by telnet backend.

Commit:
shell: add a directory for backends

This change organizes the file structure for the shell.
Files implementing backends are moved to a separate folder.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

carlescufi commented:
@jakub-uC can you rebase please?

jakub-uC commented:
@carlescufi  Some tests do not pass while they are not related to my change. What's more, locally the same tests pass correctly.
![image](https://user-images.githubusercontent.com/34125330/140508781-1c9fc12e-df57-4f7d-90b8-2f82bca4f092.png)
